README.md

```diff
@@ -114,6 +110,45 @@ history.listen((location, action) => {
 })
 ```
 
+Post-change listener functions will be not be invoked when a transition is blocked by a pre-change hook, unless the user used the built-in browser back or forward features to initiate the transition. In this case, the listener functions will be called with the same `location` and `action` as the previous invocation.
+
+#### Pre-change hook
+You can perform side effects before history state changes by registering functions with `history.before`
+
+```js
+const unsub = history.before((location, action) => {
+  location.query = queryParseFn(location.search)
+})
+```
+
+Before hooks can also be `async` functions, i.e. return a Promise, to block the state change asynchronously.
+
+```js
+const unsub = history.before(async (location, action) => {
+  await getDataAndCodeForLocation(location)
```

Comment:
What is the user looking at while we wait for `getDataAndCodeForLocation` to complete? The URL has changed, but they're still looking at the old page because history hasn't notified anyone. :/

Reply:
Exactly. The user still has the ability to see the current page, with all of its loaded data and information, while the new one loads in the background. The view only switches once the data is available and the page can be shown. The user can be shown some kind of progress indicator while that's happening.

This design can be desirable on slower networks where there's a significant amount of latency. Instead of having my users stare at a blank page with spinners or placeholder UI, I can allow them to keep looking at the data they already have. This is the behavior of Chrome and most browsers, I believe. Clicking on a link for a new document will leave the previous document there until any parts of the new document have been received.

Reply:
> The user can be shown some kind of progress indicator while that's happening.

No, you won't be able to do this because you won't know the location has changed. The user could be typing into the address bar or clicking the back button (i.e. *not* clicking on one of your links, and therefore out of your control), in which case you won't be notified the location has changed at all. Imagine someone hitting the back button and then waiting a few seconds before anything happens. They would think the site is broken.

It's not history's job to lie to you and pretend the location hasn't changed when it has. This doesn't mean that you're not free to show the user some old UI while you load some data. You absolutely can. Just look at the way React's transition group works. It gets new props (children), and gradually fades the old ones out. But you can't fade the new ones in unless you know what they are first! You can think about this library in the same way. If you're making a transition from one location to another, you will probably need to know what the destination is in order to do anything meaningful.

I hope that made sense. I'm sorry, but I don't think this is the right approach.

Reply:
You would know the location has started to change from the function that was registered with `before`. That function can make the calls it needs to change state to show a loading indicator or do whatever else.

> If you're making a transition from one location to another, you will probably need to know what the destination is in order to do anything meaningful.

You do know the destination. It's the argument given to the functions registered with `history.before`.